packer/resources/features/mongo-opsmanager/scripts/snapshot_backup.rb

```diff
@@ -0,0 +1,182 @@
+  AwsHelper::EC2::get_instances(tags)
+end
+
+def check_if_snapshot_new(snapshot_id)
+  `grep #{snapshot_id} /tmp/last_snapshot_downloaded.txt`.length == 0
+end
+
+def download_import_team_keys
+  keyring_document_key = 'flexbackupkeys.gpg'
+  logger.info("Downloading document #{keyring_document_key} from bucket #{@options.keys_bucket}")
+  s3 = Aws::S3::Client.new
+  s3.get_object({bucket: @options.keys_bucket, key: keyring_document_key}, target: '/tmp/flexbackupkeys.gpg')
+  `gpg --homedir /home/mongo-backup/ --import /tmp/flexbackupkeys.gpg`
+end
+
+def generate_gpg_command(filename)
+  keys = `gpg --homedir /home/mongo-backup/ --list-keys | grep uid`
+  key_uid_list = keys.split('\n')
+  emails = key_uid_list.map{|uid| uid.split('<')[1].tr('>', '').tr("\n", '')}
+  emails_as_args = emails.map{|email| "-r #{email}"}.join(' ')
+  encrypt_command =  "gpg --homedir /home/mongo-backup/ -e #{emails_as_args} --trust-model always -o /backup/#{filename}"
+  logger.info("GPG command: #{encrypt_command}")
+  encrypt_command
+end
+
+def download_encrypt_backup(download_link)
+  # download the file
+  file_name=download_link.split('/')[-1] + '.gpg'
+  gpg_command = generate_gpg_command(file_name)
+  download_encrypt_command = "curl #{download_link} | #{gpg_command}"
+  logger.info("Downloading and encrypting using command #{download_encrypt_command}.")
+  `#{download_encrypt_command}`
+  file_name
+end
+
+def upload_to_s3(file_name)
+  backup_location = "/backup/#{file_name}"
+  time = Time.now
+  year_month_day = "#{time.year}-#{time.month}-#{time.day}"
+  year_month_day_time = "#{year_month_day}-#{time.hour}:#{time.min}#{time.zone}"
+  key = "#{year_month_day}/#{year_month_day_time}-#{file_name}"
+
+  logger.info("Uploading #{backup_location} to bucket #{@options.backup_bucket} with key #{key}")
+  s3 = Aws::S3::Resource.new
+  object = s3.bucket(@options.backup_bucket).object(key)
+  object.upload_file(backup_location)
+  `rm #{backup_location}`
+end
+
+## main
+exit if __FILE__ != $0
+
+# use instance profile (when on instance)
+Aws.config[:credentials] = Aws::InstanceProfileCredentials.new
+Aws.config[:region] = AwsHelper::Metadata::region
+
+@options = parse_options(ARGV)
+
+# Set up global logger
+Util::SingletonLogger.instance.init_syslog(
+    ident = 'add_self_to_replset',
+    facility = Syslog::LOG_LOCAL1,
+    quiet_mode = @options.quiet_mode
+)
+
+logger.info('MongoDB: Backup snapshot...')
+
+tags = AwsHelper::InstanceData::get_tags
+rs_key = [tags['Stack'], 'db', tags['Stage']].join('-')
+
+replica_set_config = MongoDB::ReplicaSetConfig.new(nil, rs_key)
+
+locksmith = Locksmith::DynamoDB.new(
+    lock_table_name = 'mongo-initialisation',
+    max_attempts = 240,
+    lock_retry_time = 10,
+    ttl = 3600
+)
+
+locksmith.lock(replica_set_config.key) do
+  ops_manager_config = replica_set_config.ops_manager_data
+  @ops_manager = MongoDB::OpsManager.new(MongoDB::OpsManagerAPI.new(ops_manager_config))
+end
```

Comment:
I'm not sure you need to wrap this in a lock as you are
 - reading and not updating
 - not carrying out activities that need to be done exclusively (i.e. making writes to the configuration API)